File: app/src/main/java/byd/cxkcxkckx/watchdog/HangWatcher.java
package byd.cxkcxkckx.watchdog;

import android.content.SharedPreferences;
import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

/**
 * ANR-like watcher that detects main thread stalls (>5s) and persists a hang report
 * for viewing on the next app launch.
 */
public class HangWatcher {
    private static final long CHECK_INTERVAL_MS = 1000; // check every second
    private static final long THRESHOLD_MS = 5000; // consider hung if blocked >5s
    private static final String TAG = "HangWatcher";

    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final Context context;
    private long lastTick;
    private volatile boolean triggered = false;

    public HangWatcher(Context ctx) {
        this.context = ctx.getApplicationContext();
    }

    public void start() {
        lastTick = System.currentTimeMillis();
        mainHandler.post(tickRunnable);
        new Thread(monitorRunnable, "HangWatcher").start();
        Log.d(TAG, "HangWatcher started");
    }

    private final Runnable tickRunnable = new Runnable() {
        @Override
        public void run() {
            lastTick = System.currentTimeMillis();
            mainHandler.postDelayed(this, CHECK_INTERVAL_MS);
        }
    };

    private final Runnable monitorRunnable = new Runnable() {
        @Override
        public void run() {
            while (!triggered) {
                long now = System.currentTimeMillis();
                if (now - lastTick > THRESHOLD_MS) {
                    triggered = true;
                    onHangDetected(now - lastTick);
                    break;
                }
                try {
                    Thread.sleep(CHECK_INTERVAL_MS);
                } catch (InterruptedException ignored) {
                }
            }
        }
    };

    private void onHangDetected(long durationMs) {
        Log.w(TAG, "Main thread hang detected: " + durationMs + " ms");
        final String report = String.format("应用卡死 %d ms\n线程堆栈:\n%s", durationMs, buildThreadDump());
        
        // Save report synchronously (will be shown on next launch)
        try {
            SharedPreferences sp = context.getSharedPreferences(CrashApplication.PREF_NAME, Context.MODE_PRIVATE);
            Log.d(TAG, "Saving hang report to SharedPreferences");
            sp.edit()
                    .putString(CrashApplication.KEY_CRASH_REPORT, report)
                    .commit();
            Log.d(TAG, "Hang report saved successfully");
        } catch (Exception ex) {
            Log.e(TAG, "Failed to save hang report", ex);
        }
        
        // No dialog is shown; report will appear on next launch
    }

    private String buildThreadDump() {
        StringBuilder sb = new StringBuilder();
        java.util.Map<Thread, StackTraceElement[]> all = Thread.getAllStackTraces();
        for (java.util.Map.Entry<Thread, StackTraceElement[]> e : all.entrySet()) {
            sb.append("Thread: ").append(e.getKey().getName()).append("\n");
            for (StackTraceElement el : e.getValue()) {
                sb.append("  at ").append(el.toString()).append("\n");
            }
        }
        return sb.toString();
    }
}

File: app/src/main/java/momoi/mod/qqpro/watchdog/HangWatcher.kt
package momoi.mod.qqpro.watchdog

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import momoi.mod.qqpro.util.Utils
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Detects main-thread stalls. A ticker reposts itself on the main looper every second; a daemon
 * thread checks how long ago the last tick ran. If the main thread hasn't ticked for
 * [THRESHOLD_MS], it's considered hung: a report (with a thread dump) is saved and the
 * (separate-process) [CrashReportActivity] is shown. Unlike a crash we do NOT kill the process —
 * a hang may be transient — and we re-arm once the main thread recovers, so a single stall
 * reports only once.
 */
class HangWatcher(private val context: Context) {
    private val mainHandler = Handler(Looper.getMainLooper())

    @Volatile
    private var lastTick = 0L

    @Volatile
    private var reported = false

    private val ticker = object : Runnable {
        override fun run() {
            lastTick = System.currentTimeMillis()
            mainHandler.postDelayed(this, CHECK_INTERVAL_MS)
        }
    }

    fun start() {
        lastTick = System.currentTimeMillis()
        mainHandler.post(ticker)
        Thread({ monitor() }, "QQPro-HangWatcher").apply {
            isDaemon = true
            start()
        }
        Utils.log("Watchdog: HangWatcher started (threshold ${THRESHOLD_MS}ms)")
    }

    private fun monitor() {
        while (true) {
            try {
                Thread.sleep(CHECK_INTERVAL_MS)
            } catch (_: InterruptedException) {
            }
            val stalled = System.currentTimeMillis() - lastTick
            if (stalled > THRESHOLD_MS) {
                if (!reported) {
                    reported = true
                    onHang(stalled)
                }
            } else {
                // Main thread recovered → re-arm for the next hang.
                reported = false
            }
        }
    }

    private fun onHang(durationMs: Long) {
        Log.w("Watchdog", "main thread hang: ${durationMs}ms")
        Utils.log("Watchdog: main thread hang ${durationMs}ms")
        val report = buildString {
            append("应用卡死 $durationMs ms\n")
            append("时间: ${timestamp()}\n\n")
            append("主线程堆栈:\n")
            append(mainThreadStack())
            append("\n\n全部线程堆栈:\n")
            append(allThreadStacks())
        }
        Watchdog.report(context, Watchdog.KIND_HANG, report)
    }

    private fun mainThreadStack(): String =
        Looper.getMainLooper().thread.stackTrace.joinToString("\n") { "  at $it" }

    private fun allThreadStacks(): String = buildString {
        for ((thread, frames) in Thread.getAllStackTraces()) {
            append("Thread: ${thread.name}\n")
            for (f in frames) append("  at $f\n")
        }
    }

    private fun timestamp(): String =
        SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault()).format(Date())

    companion object {
        private const val CHECK_INTERVAL_MS = 1000L

        // 8s: high enough to avoid false positives during heavy work (image decode, cold start)
        // on a slow watch, while still catching real freezes.
        private const val THRESHOLD_MS = 8000L
    }
}
